### DISCUSSION TITLE
'ram'-Namespace problem(/rsm:ExchangedDocument/ram:IncludedNote) InvoiceDescriptor23CIIReader

### DISCUSSION THREAD

**Discussion by DicomSiarheiVishneuski** [stephanstapel/ZUGFeRD-csharp#473] [General]
I have one error with Message: "Namespace prefix 'ram' is not defined." in ExchangedDocument->IncludedNote(/rsm:ExchangedDocument/ram:IncludedNote) area by read.

**Comment by stephanstapel** [comment]
Can you share the invoice file and your code?

**Comment by DicomSiarheiVishneuski** [comment]
[xrechnung-Test.xml.txt](https://github.com/user-attachments/files/17980445/xrechnung-Test.xml.txt)
xrechnung-info-file
listing:
var filePath = @"d:\temp\xrechnung-Test.xml.txt";
using (FileStream stream = new FileStream(attachmentPath, FileMode.Open))
{
    zugferdList.Add(InvoiceDescriptor.Load(stream));

    stream.Flush();
    stream.Close();
}

**Comment by DicomSiarheiVishneuski** [comment]
I use the in the project: Framework 4.7, but with the version of the core 5, I got the same error

**Comment by stephanstapel** [comment]
thanks.
The prefixes ram and rsm should be used from your supplier:

https://www.zugferd-community.net/de/infothek/technical_xml_regarding_fixed_namespace_prefixes_rsm_ram_

**Comment by DicomSiarheiVishneuski** [comment]
link not work:
![image](https://github.com/user-attachments/assets/bad33039-ce30-41a8-9b3d-37b6cd08c44c)

**Comment by stephanstapel** [comment]
The first link here:

https://www.google.de/search?q=rsm+ram+zugferd+infothek&sca_esv=d1e3e92ad7152243&sxsrf=ADLYWIIFbRKtXTtuPPOXwd6pBkP_m0SVUw%3A1733155808656&ei=4NtNZ-3KJ6i8xc8PoLyi4A8&ved=0ahUKEwjtusLCvImKAxUoXvEDHSCeCPwQ4dUDCA8&uact=5&oq=rsm+ram+zugferd+infothek&gs_lp=Egxnd3Mtd2l6LXNlcnAiGHJzbSByYW0genVnZmVyZCBpbmZvdGhlazIFECEYoAFI6BdQnAVYlhZwBHgBkAEAmAGQAaAB9gaqAQQxMS4xuAEDyAEA-AEBmAIOoALSBcICChAAGLADGNYEGEfCAgcQIRigARgKwgIEECEYFZgDAIgGAZAGCJIHAjE0oAfGHg&sclient=gws-wiz-serp

**Comment by DicomSiarheiVishneuski** [comment]
yes, if namespaces "rsm, ram" are mandatory, and in your code they are visible among the names, why are they not in the list of namespace-prefixes?

**Comment by DicomSiarheiVishneuski** [comment]
I received this invoice and cannot load it using the library. What should I do? can you me help, please

**Comment by DicomSiarheiVishneuski** [comment]
i will tolk, that a function 'XmlNamespaceManager nsmgr = _GenerateNamespaceManagerFromNode(doc.DocumentElement);' is false , or the  loop foreach (XmlNode node in doc.SelectNodes("//rsm:ExchangedDocument/ram:IncludedNote", nsmgr)) in class "InvoiceDescriptor23CIIReader".

**Comment by DicomSiarheiVishneuski** [comment]
here is a file in which a namespace error occurs: [xrechnung-Test.xml.txt](https://github.com/user-attachments/files/17992678/xrechnung-Test.xml.txt)

**Comment by DicomSiarheiVishneuski** [comment]
if you add namespaces xmlns:a="urn:un:unece:uncefact:data:standard:QualifiedDataType:100" xmlns:rsm="urn:un:unece:uncefact:data:standard:CrossIndustryInvoice:100" xmlns:qdt="urn:un:unece:uncefact:data:standard:QualifiedDataType:100" xmlns:ram="urn:un:unece:uncefact:data:standard:ReusableAggregateBusinessInformationEntity:100" xmlns:xs="http://www.w3.org/2001/XMLSchema" xmlns:udt="urn:un:unece:uncefact:data:standard:UnqualifiedDataType:100" in <CrossIndustryInvoice ....> , then everything is loaded

**Comment by RBITSRobin** [comment]
How do we handle situations where a system uses its own namespaces?

I have an example here that was generated from SAP.
The used namespaces are:
xmlns:n0="urn:un:unece:uncefact:data:standard:CrossIndustryInvoice:100" 
xmlns:prx="urn:sap.com:proxy:Q75:/1SAI/TASA78ABB6156583BE4C420:755" 
xmlns:n1="urn:un:unece:uncefact:data:standard:QualifiedDataType:100" 
xmlns:n2="urn:un:unece:uncefact:data:standard:ReusableAggregateBusinessInformationEntity:100" 
xmlns:n3="urn:un:unece:uncefact:data:standard:UnqualifiedDataType:100"

when i add the ones provided by @DicomSiarheiVishneuski then the document will be processed. But i don´t want to modify incoming invoices in production ;-)

Is the conclusion that the invoice is incorrect and the manufacturer is responsible for issuing it correctly?

**Comment by RBITSRobin** [comment]
<n2:IssueDateTime>
   <n3:DateTimeString format="102">20251024</n3:DateTimeString>
  </n2:IssueDateTime>
  
  when i use the InvoiceDescriptor invoice = InvoiceDescriptor.Load(pathToXML); function InvoiceDate will be null.
  I think it´s because the namespace used here is "n3" instead of "qdt".

**Comment by stephanstapel** [comment]
Thanks for sending this issue. Can you share an invoice (email address on the readme page)?
